### PR TITLE
Add config entry for unsafe retries on container runner [ONPREM-1081]

### PR DESCRIPTION
**Pull request by christian-stephen**
# Description
A brief description of the changes.

Add a config reference entry for setting up and explaining unsafe retries. Design doc: https://circleci.atlassian.net/l/cp/eDoD37SS

# Reasons
A link to a GitHub and/or JIRA issue (if applicable).
Otherwise, a brief sentence about why you made these changes.

https://circleci.atlassian.net/browse/ONPREM-1081

# Content Checklist
Please follow our style when contributing to CircleCI docs. Our style guide is here: [https://circleci.com/docs/style/style-guide-overview](https://circleci.com/docs/style/style-guide-overview).

Please take a moment to check through the following items when submitting your PR (this is just a guide so will not be relevant for all PRs) 😸:

- [ ] Break up walls of text by adding paragraph breaks.
- [ ] Consider if the content could benefit from more structure, such as lists or tables, to make it easier to consume.
- [ ] Keep the title between 20 and 70 characters.
- [ ] Consider whether the content would benefit from more subsections (h2-h6 headings) to make it easier to consume.
- [ ] Check all headings h1-h6 are in sentence case (only first letter is capitalized).
- [ ] Is there a "Next steps" section at the end of the page giving the reader a clear path to what to read next?
- [ ] Include relevant backlinks to other CircleCI docs/pages.

### CODE REVIEW

File: jekyll/_cci2/container-runner.adoc [hunk 1]
```diff
@@ -95,6 +95,51 @@ agent:
           - name: "other"
 ```
 
+[#unsafe-retries]
+=== Unsafe retries
+
+Unsafe retries is a feature that enables container runner to automatically rerun tasks if there are unexpected interruptions during their execution. These disruptions could be due to network connectivity issues, the underlying node shutting down, or other unpredictable causes. This feature can also be used when scheduling workloads on spot instances, which often come with cost-saving benefits at the risk of pod preemptions with many Kubernetes providers.
```
Comment (ryanwohara) [inline jekyll/_cci2/container-runner.adoc:101]: The singular/plural start of the paragraph feels a little awkward

**Review by atulsingh0 — APPROVED** [review]

File: jekyll/_cci2/container-runner.adoc [hunk 2]
```diff
@@ -95,6 +95,51 @@ agent:
...
 ```
 
+[#unsafe-retries]
+=== Unsafe retries
+
+Unsafe retries enable container runner to automatically rerun tasks if there are unexpected interruptions during their execution. These disruptions could be due to network connectivity issues, the underlying node shutting down, or other unpredictable causes. This feature can also be used when scheduling workloads on spot instances, which often come with cost-saving benefits at the risk of pod preemptions with many Kubernetes providers.
+
+[CAUTION]
+====
+This feature is called “unsafe retries” for a reason. Unlike automatic retries on startup, retrying tasks during runtime can be risky. This is because tasks can have arbitrary steps that produce side effects which are not idempotent or stateless. It is important to use unsafe retries knowing the impact it may have on jobs and workflows. Use this feature with jobs and workflows that are fault-tolerant.
```
Comment (jrahme-cci) [inline jekyll/_cci2/container-runner.adoc:105]: Instead of noting the impact it will have on jobs and workflows I think we would do well to warn about knowing the impact reran steps may have on external environments and resources such as production environments or databases. 

I'd also change "Use this feature with jobs and workflows that are fault tolerant" to "Use this feature with care, knowing the risks of rerunning jobs and workflows that may or may not be idempotent"

File: jekyll/_cci2/container-runner.adoc [hunk 3]
```diff
@@ -95,6 +95,51 @@ agent:
           - name: "other"
 ```
 
+[#unsafe-retries]
+=== Unsafe retries
+
+Unsafe retries enable container runner to automatically rerun tasks if there are unexpected interruptions during their execution. These disruptions could be due to network connectivity issues, the underlying node shutting down, or other unpredictable causes. This feature can also be used when scheduling workloads on spot instances, which often come with cost-saving benefits at the risk of pod preemptions with many Kubernetes providers.
```
Comment (jrahme-cci) [inline jekyll/_cci2/container-runner.adoc:101]: Worth noting that errors surfaced in our UI as infrastructure failures should be expected to trigger an unsafe retry

File: jekyll/_cci2/container-runner.adoc [hunk 4]
```diff
@@ -95,6 +95,51 @@ agent:
...
+
+[CAUTION]
+====
+This feature is called “unsafe retries” for a reason. Unlike automatic retries on startup, retrying tasks during runtime can be risky. This is because tasks can have arbitrary steps that produce side effects which are not idempotent or stateless. It is important to use unsafe retries knowing the impact it may have on jobs and workflows. Use this feature with jobs and workflows that are fault-tolerant.
+====
+
+Here is how unsafe retries work with container runner:
+
+1. If a pod fails or gets evicted during runtime, container runner will "un-claim", or release the task.
```
Comment (jrahme-cci) [inline jekyll/_cci2/container-runner.adoc:110]: un-claim may be exposing an implementation detail, I think we're okay to just say "release the task"

File: jekyll/_cci2/container-runner.adoc [hunk 5]
```diff
@@ -95,6 +95,51 @@ agent:
...
+2. All associated resources of the task are cleaned up.
```
Comment (jrahme-cci) [inline jekyll/_cci2/container-runner.adoc:111]: s/All associated resources/All CircleCI Self Hosted Runner cluster resources/

This could give the impression that we'd somehow cleanup generated data or state external to the task management.

File: jekyll/_cci2/container-runner.adoc [hunk 6]
```diff
@@ -95,6 +95,51 @@ agent:
...
+3. The released task then becomes available for reclaim by any container runner instance configured for the same resource class.
+4. Any previously ran steps will be reset, and the task will restart from scratch.
```
Comment (jrahme-cci) [inline jekyll/_cci2/container-runner.adoc:113]: The wording of "reset" here may also give the impression that we're undoing side effects somehow. Maybe simplify to "The task will be started from scratch, including previously ran steps".

I'm not sure about the usage of "task" vs "job" here. We don't really expose the concept of a task to customers

**Comment by jrahme-cci** [comment]
I'm not sure how much we should be sticking to the "Self Hosted Runner" branding in our docs. @nathanfishcircleci or @rosieyohannan any thoughts on the use of "container runner" here vs "self hosted container runner"

File: jekyll/_cci2/container-runner.adoc (hunk 4 above)
Comment (christian-stephen) [reply jekyll/_cci2/container-runner.adoc:110]: Disagree. We talk about claiming and tasks in the docs [here](https://circleci.com/docs/container-runner/#resource-class-configuration-custom-pod). I think mentioning unclaiming here emphasizes that it may be claimed again by another instance of container agent.

File: jekyll/_cci2/container-runner.adoc (hunk 2 above)
Comment (christian-stephen) [reply jekyll/_cci2/container-runner.adoc:105]: This was lifted from Product wording/warning on the feature when previously presented to customers, but I can reword this to that.

File: jekyll/_cci2/container-runner.adoc (hunk 6 above)
Comment (christian-stephen) [reply jekyll/_cci2/container-runner.adoc:113]: We do expose the concept of tasks to customers: https://circleci.com/docs/faq/#what-is-a-CircleCI-task-vs-a-job

Also within this same document: https://circleci.com/docs/container-runner/#resource-class-configuration-custom-pod

**Comment by christian-stephen** [comment]
> I'm not sure how much we should be sticking to the "Self Hosted Runner" branding in our docs. @nathanfishcircleci or @rosieyohannan any thoughts on the use of "container runner" here vs "self hosted container runner"

I feel that discussion is a bit out of scope for this PR. We use "container runner" in other places in this document. Maybe open a separate discussion thread?

**Review by imlogang — APPROVED** [review]

File: jekyll/_cci2/container-runner.adoc [hunk 7]
```diff
@@ -95,6 +95,53 @@ agent:
           - name: "other"
 ```
 
+[#unsafe-retries]
+=== Unsafe retries
+
+Unsafe retries enable container runner to automatically rerun tasks that are unexpectedly interrupted during their execution. These disruptions could be due to network connectivity issues, the underlying node shutting down, or other unpredictable causes. Any job failure that would normally be displayed in the CircleCI web app as an infrastructure fail should be expected to trigger an unsafe retry when enabled.
```
Comment (rosieyohannan) [inline jekyll/_cci2/container-runner.adoc:101]:
```suggestion
Unsafe retries enable container runner to automatically rerun tasks that are unexpectedly interrupted during their execution. These disruptions could be due to network connectivity issues, the underlying node shutting down, or other unpredictable causes. Any job failure that would be displayed in the CircleCI web app as an infrastructure fail should be expected to trigger an unsafe retry when enabled.
```

File: jekyll/_cci2/container-runner.adoc [hunk 8]
```diff
@@ -95,6 +95,53 @@ agent:
...
+
+This feature is also useful when scheduling workloads on spot instances, which often come with cost-saving benefits at the risk of pod preemptions with many Kubernetes providers.
```
Comment (rosieyohannan) [inline jekyll/_cci2/container-runner.adoc:103]:
```suggestion
Unsafe retries is useful when scheduling workloads on spot instances, which often come with cost-saving benefits at the risk of pod preemptions with many Kubernetes providers.
```

File: jekyll/_cci2/container-runner.adoc [hunk 9]
```diff
@@ -95,6 +95,53 @@ agent:
...
+
+[CAUTION]
+====
+This feature is called “unsafe retries” for a reason. Unlike automatic retries on startup, retrying tasks during runtime can be risky. This is because tasks can have arbitrary steps that produce external side effects which are not idempotent or stateless. This includes steps that could impact production environments or databases. Use this feature with care, knowing the risks of rerunning jobs and workflows that may or may not be idempotent.
+====
+
+Here is how unsafe retries work with container runner:
```
Comment (rosieyohannan) [inline jekyll/_cci2/container-runner.adoc:110]:
```suggestion
The following sequence shows how unsafe retires work:
```

File: jekyll/_cci2/container-runner.adoc [hunk 10]
```diff
@@ -95,6 +95,53 @@ agent:
...
+
+1. If a pod fails or gets evicted during runtime, container runner will release the task.
+2. All resources managed by container runner for the task, such as the Kubernetes pod and secret, are cleaned up and deleted.
+3. The released task then becomes available for reclaim by any container runner instance configured for the same resource class.
+4. Once reclaimed, the task is restarted completely from scratch, including previously run steps.
+5. A task can be retried up to 3 times before it is deemed to have permanently failed.
+
+To enable unsafe retries, you need to set the `enableUnsafeRetries` flag in the <<resource-class-configuration-custom-pod,resource class configuration>> for each resource class. Here is an example where two resource classes are defined - one with unsafe retries enabled for spot instances and the other one where it is not enabled:
```
Comment (rosieyohannan) [inline jekyll/_cci2/container-runner.adoc:118]:
```suggestion
To enable unsafe retries, set the `enableUnsafeRetries` flag in the <<resource-class-configuration-custom-pod,resource class configuration>> for each resource class. The following example shows two resource class definitions. Unsafe retries is enabled for the first, for spot instances, but not for the second resource class:
```

**Review by rosieyohannan — APPROVED** [review]

**Review by jrahme-cci — APPROVED** [review]
